### PR TITLE
support pre stop hook.

### PR DESCRIPTION
**Pull request by zonghaishang**
### Issues associated with this PR

Your PR should present related issues you want to solve.

### Solutions
You should show your solutions about the issues in your PR, including the overall solutions, details and the changes. At this time, Chinese is allowed to describe these.

### UT result
Unit Test is needed if the code is changed, your unit test should cover boundary cases, corner cases, and some exceptional cases. And you need to show the UT result.

### Benchmark
If your code involves the processing of every request, you should give the Benchmark Result.

### Code Style
+ Make sure `Goimports` has run
+ Show `Golint` result

### CODE REVIEW

**Comment by zonghaishang** [comment]
@cch123 @nejisama please take a look

File: pkg/server/handler.go [hunk 1]
```diff
@@ -488,8 +489,91 @@ func (al *activeListener) OnNewConnection(ctx context.Context, conn api.Connecti
...
+			}
+		}
+	}
+
+	return activeStream
+}
+
 func (al *activeListener) OnClose() {}
 
+func (al *activeListener) PreStopHook(ctx context.Context) func() error {
```
Comment (cch123) [inline pkg/server/handler.go:536]: explain why this function is needed and when to use it
